src/hb-ot-hmtx-table.hh

```diff
@@ -130,33 +132,59 @@ struct hmtxvmtx
 
     accelerator_t _mtx (c->plan->source);
     unsigned num_long_metrics;
+    const hb_hashmap_t<unsigned, hb_pair_t<unsigned, int>> *mtx_map = get_mtx_map (c->plan);
     {
       /* Determine num_long_metrics to encode. */
       auto& plan = c->plan;
+
       num_long_metrics = plan->num_output_glyphs ();
-      hb_codepoint_t old_gid = 0;
-      unsigned int last_advance = plan->old_gid_for_new_gid (num_long_metrics - 1, &old_gid) ? _mtx.get_advance_without_var_unscaled (old_gid) : 0;
+      unsigned int last_advance = get_advance (plan, mtx_map, num_long_metrics - 1, _mtx);
       while (num_long_metrics > 1 &&
-	     last_advance == (plan->old_gid_for_new_gid (num_long_metrics - 2, &old_gid) ? _mtx.get_advance_without_var_unscaled (old_gid) : 0))
+	     last_advance == get_advance (plan, mtx_map, num_long_metrics - 2, _mtx))
       {
 	num_long_metrics--;
       }
     }
 
-    auto it =
-    + hb_range (c->plan->num_output_glyphs ())
-    | hb_map ([c, &_mtx] (unsigned _)
-	      {
-		hb_codepoint_t old_gid;
-		if (!c->plan->old_gid_for_new_gid (_, &old_gid))
-		  return hb_pair (0u, 0);
-		int lsb = 0;
-		(void) _mtx.get_leading_bearing_without_var_unscaled (old_gid, &lsb);
-		return hb_pair (_mtx.get_advance_without_var_unscaled (old_gid), +lsb);
-	      })
-    ;
-
-    table_prime->serialize (c->serializer, it, num_long_metrics);
+    if (mtx_map->is_empty ())
+    {
+      auto it =
+      + hb_range (c->plan->num_output_glyphs ())
+      | hb_map ([c, &_mtx] (unsigned _)
+		{
+		  hb_codepoint_t old_gid;
+		  if (!c->plan->old_gid_for_new_gid (_, &old_gid))
+		    return hb_pair (0u, 0);
+		  int lsb = 0;
+		  (void) _mtx.get_leading_bearing_without_var_unscaled (old_gid, &lsb);
+		  return hb_pair (_mtx.get_advance_without_var_unscaled (old_gid), +lsb);
+		})
+      ;
+
+      table_prime->serialize (c->serializer, it, num_long_metrics);
+    }
+    else
```

Comment:
Why this code duplication? Can we avoid it?

Reply:
The code is basically like:
if (a)
{
  auto it = {iterator generation code}
  serialize (it); 
}
else
{
  auto it = {iterator generation code}
  serialize (it);
}

I tried to avoid duplicated code for serialize() call, but it seems there's not an easy way. Because the type of iterator will be different in either branch. So the serialize call also uses different argument types, thus I'll need two distinct calls

Reply:
done.